### ISSUE TITLE
"no-invalid-this" gives false-negative result

### ISSUE THREAD

**Issue by vitalymak** [eslint/eslint#9126]
* **ESLint Version:** 4.4.1
* **Node Version:** 8.2.1
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**What did you do? Please include the actual source code causing the issue.**

```sh
../node_modules/.bin/eslint --no-eslintrc no-invalid-this-test.js 
```

**no-invalid-this-test.js** file contents:
```js
/*eslint no-invalid-this: "error"*/
/*eslint-env es6*/

let o = {
    name: "O name",
    f: function() {
        console.log(this.name)
    }
};

o.f = ((func, arg) => function() {
    console.log('Executing wrapper with arg ' + arg);
    func.call(this);
})(o.f, 'blah');

o.f();
```

**What did you expect to happen?**
No error.

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
/Users/vitaly/Work/n/build/no-invalid-this-test.js
  13:15  error  Unexpected 'this'  no-invalid-this

✖ 1 problem (1 error, 0 warnings)
```

**Comment by not-an-aardvark** [comment]
Thanks for the report. However, this is working as intended -- the [rule documentation](https://eslint.org/docs/rules/no-invalid-this#rule-details) defines what it means for a use of `this` to be "valid", and your code does not meet that definition.

**Comment by not-an-aardvark** [comment]
Hmm, it looks like there's some special logic being used for `return` statements to check for method assignments ([relevant code](https://github.com/eslint/eslint/blob/ffa021e7696b24722cbbef306d494dd64d7b5215/lib/ast-utils.js#L623-L629)). We should probably update that to check for arrow function returns as well.
